### PR TITLE
add mask support for flash attention

### PR DESCRIPTION
**Pull request by Cjkkkk**
* flash attention only support causal mask right now, cuDNN also supports take in a mask directly. Add mask support for flash attention.

### CODE REVIEW

**Comment by Cjkkkk** [comment]
@akuegel Hey Adrian, could you take a look at this PR?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 1]
```diff
@@ -736,7 +738,6 @@ MatchFwdResult MatchBmm1ScaleBiasMaskSoftmaxDropoutBmm2(
       matched_result.has_match = false;
       return matched_result;
     }
-    matched_result.is_causal_mask |= IsCausalMaskPattern(mask);
```
Comment (akuegel) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:739]: Why do we need to remove this line?

File: xla/service/gpu/tests/gpu_fused_mha_test.cc [hunk 2]
```diff
@@ -2709,6 +2709,136 @@ class FlashAttentionBMMScaleBiasSoftmaxBMM : public MultiHeadedAttentionTest {
...
+    auto lhs_bmm1_literal =
+        GetInput4DLiteral<T>({2, 6, 1024, 64}, {3, 2, 1, 0});
+    auto rhs_bmm1_literal =
+        GetInput4DLiteral<T>({2, 6, 64, 1024}, {3, 2, 1, 0});
+    auto rhs_bmm2_literal =
+        GetInput4DLiteral<T>({2, 6, 1024, 64}, {3, 2, 1, 0});
+    auto mask_literal = GetMask4DLiteral({2, 6, 1024, 1024}, {3, 2, 1, 0});
+    auto do_literal = GetInput4DLiteral<T>({2, 6, 1024, 64}, {3, 2, 1, 0});
+    std::string hlo_string = "";
+    hlo_string =
```
Comment (akuegel) [inline xla/service/gpu/tests/gpu_fused_mha_test.cc:2834]: Why not just directly assign?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 1 above)
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:739]: Current CausalMask pattern match only match a causal mask that is added to the bmm1 output. Would be better to do this in a separate PR.

File: xla/service/gpu/tests/gpu_fused_mha_test.cc (hunk 2 above)
Comment (Cjkkkk) [reply xla/service/gpu/tests/gpu_fused_mha_test.cc:2834]: Fixed.

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 1 above)
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:739]: Also require to rewrite some parts of rewriter logic to consider cases where bias and mask could both be causalMask.

Comment (akuegel) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:739]: If I understand correctly, not removing this line would currently be an issue and you intend to add some additional logic in another PR? Sounds good.

**Review by akuegel — APPROVED** [review]

**Comment by Cjkkkk** [comment]
@akuegel Not moving this line will not cause an issue as it will just not match. But remove this makes it clear that mask as causal mask is not supported right now.

**Comment by akuegel** [comment]
> @akuegel Not moving this line will not cause an issue as it will just not match. But remove this makes it clear that mask as causal mask is not supported right now.

Makes sense, thanks for the explanation.

**Review by akuegel — APPROVED** [review]

**Comment by Cjkkkk** [comment]
@akuegel Seems like the Google internal checks FAILED?

**Comment by akuegel** [comment]
> @akuegel Seems like the Google internal checks FAILED?

That is not a real failure, it complained about a missing TESTED= tag. I have added it. Nothing for you to worry about :)
But the internal approval will probably only happen during US working hours.

**Pull request closed without merging**